sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.catalyst.expressions.AttributeReference
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, Statistics}
+import org.apache.spark.sql.sources.v2.reader.DataSourceV2Reader
+import org.apache.spark.sql.sources.v2.reader.upward.StatisticsSupport
+
+case class DataSourceV2Relation(
+    output: Seq[AttributeReference],
+    reader: DataSourceV2Reader) extends LeafNode {
+
+  override def computeStats(): Statistics = reader match {
+    case r: StatisticsSupport => Statistics(sizeInBytes = r.getStatistics.sizeInBytes())
+    case _ => Statistics(sizeInBytes = conf.defaultSizeInBytes)
+  }
+}
+
+object DataSourceV2Relation {
+  def apply(reader: DataSourceV2Reader): DataSourceV2Relation = {
+    new DataSourceV2Relation(reader.readSchema().toAttributes, reader)
```

Comment:
Is this the right schema? The docs for `readSchema` say it is the result of pushdown and projection, which doesn't seem appropriate for a `Relation`. Does relation represent a table that can be filtered and projected, or does it represent a single read? At least in the Hive read path, it's a table.

Reply:
I leave it as a [TODO](https://github.com/apache/spark/pull/19136#discussion_r137023744) as it needs some refactoring on the optimizer. For now `DataSourceV2Relation` represents a data source without any optimization: we do these optimizations during planning. This is also a problem for data source v1, and that's why we implement partition pruning as an optimizer rule instead of data source internal, because we need to update the stats.

Reply:
Are you saying that partition pruning isn't delegated to the data source in this interface?

I was just looking into how the data source should provide partition data, or at least fields that are the same for all rows in a `ReadTask`. It would be nice to have a way to pass those up instead of materializing them in each `UnsafeRow`.

Reply:
In data source V2, we will delegate partition pruning to the data source, although we need to do some refactoring to make it happen.

> I was just looking into how the data source should provide partition data, or at least fields that are the same for all rows in a `ReadTask`. It would be nice to have a way to pass those up instead of materializing them in each `UnsafeRow`.

This can be achieved by the columnar reader. Think about a data source having a data column `i` and a partition column `j`, the returned columnar batch has 2 column vectors for `i` and `j`. Column vector `i` is a normal one that contains all the values of column `i` within this batch, column vector `j` is a constant vector that only contains a single value.

Reply:
I think we should add a way to provide partition values outside of the columnar reader. It wouldn't be too difficult to add a method on `ReadTask` that returns them, then create a joined row in the scan exec. Otherwise, this requires a lot of wasted memory for a scan.

Reply:
I think users can write a special `ReadTask` to do it, but we can't save memory by doing this. When an operator(the scan operator) transfers data to another operator, the data must be `UnsafeRow`s. So even users return a joined row in `ReadTask`, Spark need to convert it to `UnsafeRow`.